### ISSUE TITLE
default_number_of_transfer_threads seems to be ignored

### ISSUE THREAD

**Issue by bh9** [irods/irods#5701]
- [x] irods_docs

---

### iRODS Version, OS and Version
4.2.7 Ubuntu 18.04 with 16.04 packages

### What did you try to do?
iput a 60MB file. 
```
bh9@farm5-head2:~$ fallocate -l 60M /tmp/bh960M
bh9@farm5-head2:~$ iput /tmp/bh960M -V
```

### Expected behavior
Since we have `default_number_of_transfer_threads: 4` on the server acting as both provider and consumer, and I have not specified a number with `-N`, I would expect 4 threads, especially given `man iput` says `By default (-N option not used) the server decides the number of threads to use.`

### Observed behavior (including steps to reproduce, if applicable)
iput uses 16 threads
```
bh9@farm5-head2:~$ iput /tmp/bh960M -V
From server: NumThreads=16, addr:irods-sanger1-ies4, port:20125, cookie=585681288
   bh960M                         60.000 MB | 0.189 sec | 16 thr | 316.924 MB/s
```

Note: anything less than 60M will use fewer than 16 threads, but never 4 since <32MB is single-threaded and 33MB is 9 threads

**Comment by mcast** [comment]
Sorry for the distraction, but what caused `-V` to give you the `From server: NumThreads=16, addr:irods-sanger1-ies4, port:20125, cookie=585681288` ?

I can't provoke that output and it would be useful for prodding bugs (or other bug-like features).

**Comment by bh9** [comment]
> Sorry for the distraction, but what caused `-V` to give you the `From server: NumThreads=16, addr:irods-sanger1-ies4, port:20125, cookie=585681288` ?  

-V is the very verbose flag for iput (and some other icommands): `man iput` says `-V     Very verbose`

**Comment by bh9** [comment]
Note, however, that it doesn't produce that information for uploads under 32MB

**Comment by mcast** [comment]
> Note, however, that it doesn't produce that information for uploads under 32MB

Thanks, that is what tricked me. `-V  Very verbose` except when there isn't much to say, such as when `-N0` or `-N1` 😞

**Comment by trel** [comment]
Please check your core.re file for acSetNumThreads.   The second parameter is the max threads to use, and has been set to 16 for some time.  We recently changed it to ‘default’ for 4.2.9, which will honor the setting in the server_config.json.

**Comment by trel** [comment]
see https://github.com/irods/irods/issues/5654

**Comment by bh9** [comment]
If we set that to 4, are we still able to use more than 4 threads if that is requested? I can see a case where we may want more threads for a 25G link

**Comment by kript** [comment]
Slightly confused here, apologies. Core.re says;

```
# 9) acSetNumThreads - Rule to set the number of threads for a data transfer
# This rule supports condition based on $rescName so that different
# policies can be set for different resources.
# Only one function can be used for this rule
#    msiSetNumThreads(sizePerThrInMb, maxNumThr, windowSize) - set the number
#      of threads and the tcp window size. The number of threads is based
#      the two input parameters
#      sizePerThrInMb - The number of threads is computed using:
#        numThreads = fileSizeInMb / sizePerThrInMb + 1
#        where sizePerThrInMb is an integer value in MBytes. It also accepts
#        the word "default" which sets sizePerThrInMb to a default value of 32
#      maxNumThr - The maximum number of threads to use. It accepts integer
#        value up to 16. It also accepts the word "default" which sets
#        maxNumThr to a default value of 4. A value of 0 means no parallel
#        I/O. This can be helpful to get around firewall issues.
#    windowSize - the tcp window size in Bytes for the parallel transfer.
#      A value of 0 or "default" means a default size of 1,048,576 Bytes.
# The msiSetNumThreads function must be present or no thread will be used
# for all transfer
# acSetNumThreads {msiSetNumThreads("16","4","default"); }
# acSetNumThreads {msiSetNumThreads("default","16","default"); }
# acSetNumThreads {ON($rescName == "macResc") {msiSetNumThreads("default","0","default"); } }
acSetNumThreads {msiSetNumThreads("default","16","default"); }
```
So as I read this the default is set to 16. If we change the 16 to default, will 4 be used _unless_ the client specifies a different number OR a different value is set in `server_config.json` (and I am not sure if the client over-rids this?).

**Comment by kript** [comment]
Also, is it possible to set this in the users `environment.json` and/or with a environment variable?

**Comment by mcast** [comment]
This is biting us again - the `load1` on one of our iRES hosts peaked at 960 and although it's coping remarkably well (data transfer rate peaked at 1.5 GiB/sec) we can probably get more performance with fewer threads.

For now & lacking a way to set the default on the server, we will choose it explicitly in the `iget` invocation.

**Comment by alanking** [comment]
Sorry for the radio silence there...

To answer @kript's question, I think you are reading that correctly. There is not an `irods_environment.json` setting nor an environment variable recognized by the iCommands which feeds a number of threads those which care about it. That's an interesting suggestion though!

You can dial back the number of threads by changing that "16" to "default" and the `server_config.json` setting should be honored. Let us know if this is not so.

**Comment by mcast** [comment]
> Sorry for the radio silence there...

No worries, we can test / could have already tested the change above on our dev zone but last time round we forgot about it... until the problem came back.

Currently we have an override on the main caller of `iget`, but it would be nicer to remove that.

**Comment by trel** [comment]
@alanking is this a bug, as described, at this point?

I feel like this is working as intended/designed, even if it's not well explained / in a digestible form...

**Comment by alanking** [comment]
I think this mainly a documentation exercise at this point, yes.

`default_number_of_transfer_threads` is defined as "The number of threads enabled when parallel transfer is invoked." in the documentation (https://docs.irods.org/4.2.11/system_overview/configuration/)

If `acSetNumThreads` invokes `msiSetNumThreads` and sets `maxNumThr` to "default", the setting will be honored. Otherwise, it will, as the name suggests, set a maximum thread count of whatever is put there, thereby overriding the default, which means, of course, that `default_number_of_transfer_threads` would have no effect.

**Comment by trel** [comment]
Right, so, marking this as documentation, and we get your sentence into the docs somewhere real smart.

**Comment by mcast** [comment]
> Right, so, marking this as documentation, and we get your sentence into the docs somewhere real smart.

This makes sense to me - we should at this point know where the effective places are to control the thread count.

For those of us who don't deal with the server configs directly, please can you include in the docs a pointer to something describing the distinction between `acSetNumThreads` and `msiSetNumThreads` ?

Thanks, m

**Comment by kript** [comment]
> To answer @kript's question, I think you are reading that correctly. There is not an `irods_environment.json` setting nor an environment variable recognized by the iCommands which feeds a number of threads those which care about it. That's an interesting suggestion though!
> 


Could we make this into an enhancement request as well as updating the documentation?

**Comment by alanking** [comment]
So the setting being proposed would be like an implied `-N` (in `iput` parlance)?

```
 -N  numThreads - the number of threads to use for the transfer. A value of
       0 means no threading. By default (-N option not used) the server 
       decides the number of threads to use.
```

**Comment by kript** [comment]
> So the setting being proposed would be like an implied `-N` (in `iput` parlance)?
> 
> ```
>  -N  numThreads - the number of threads to use for the transfer. A value of
>        0 means no threading. By default (-N option not used) the server 
>        decides the number of threads to use.
> ```

I don't think I was clear, apologies, I was thinking of this; https://github.com/irods/irods/issues/5701#issuecomment-988124083
i.e. (modified for hopeful clarity);
> An irods_environment.json setting or an environment variable recognized by the iCommands which feeds a number of threads to those which care about it. 
If that's _also_ what you were thinking about, super!

**Comment by alanking** [comment]
Sorry, that is in fact what I was thinking about, yes. :)

**Comment by korydraughn** [comment]
See https://github.com/irods/irods/issues/6199 for new client-side setting.

**Comment by alanking** [comment]
Now that I've sat down to look things over and write down how to work with the number of transfer threads, I think there may be at least one bug in this implementation. In fact, I think the original report is revealing a bug in this implementation, even though the use case is achievable through other means.

The main issue I've found is that `acSetNumThreads` is not giving us the behavior that we want, and this is because the second parameter of `msiSetNumThreads` - `maxNumThr` - is misleading at best and incorrectly implemented at worst. When `maxNumThr` is set to anything other than "default", that is the number of threads that will be used in any parallel transfer, regardless of size, client inputs, or server configurations. So, rather than acting as a *maximum* it is acting as a unilateral override of sorts.

The good news is that a variable number of threads is still configurable as long as `maxNumThr` is set to "default". The bad news is that a lot more gymnastics have to happen on the part of the administrator to set an actual maximum number of threads because `acSetNumThreads` cannot be used in this way as near as I can tell.

Next, I have found that `irods_default_number_of_transfer_threads` being set in the client environment for both a regular user and the service account seems to have no impact on the behavior whatsoever. I think that this value is perhaps a fallback used in the event that the `default_number_of_transfer_threads` is not set in the server configuration. However, our JSON schema validation would not allow a server to start without this value, so an administrator would have to very intentionally circumvent this "bumper" to see any effect. All that being said, I'm not sure that I understand the purpose of this client configuration at this time.

So in summary, I think that there could definitely stand to be more/better documentation around this. I also think that this issue reveals a limitation in our provided tooling for controlling parallel transfer threads, which I consider to be a **bug**.

**Comment by alanking** [comment]
Documentation has been added. #6199 will address the client environment limitations.
